contracts/middleware/MiddlewareRemove.sol

```diff
@@ -0,0 +1,128 @@
+    ) external returns (bytes4, BalanceDelta) {
+        if (bytes32(hookData) == OVERRIDE_BYTES) {
+            (, bytes memory returnData) = address(implementation).delegatecall(
+                abi.encodeWithSelector(this.afterRemoveLiquidity.selector, sender, key, params, delta, hookData[32:])
+            );
+            return abi.decode(returnData, (bytes4, BalanceDelta));
+        }
+        (bool success, bytes memory returnData) = address(this).delegatecall{gas: GAS_LIMIT}(
+            abi.encodeWithSelector(this._callAndEnsureValidDeltas.selector, sender, key, params, delta, hookData)
+        );
+        if (success) {
+            return (BaseHook.afterRemoveLiquidity.selector, abi.decode(returnData, (BalanceDelta)));
+        } else {
+            return (BaseHook.afterRemoveLiquidity.selector, BalanceDeltaLibrary.ZERO_DELTA);
+        }
+    }
+
+    function _callAndEnsureValidDeltas(
+        address sender,
+        PoolKey calldata key,
+        IPoolManager.ModifyLiquidityParams calldata params,
+        BalanceDelta delta,
+        bytes calldata hookData
+    ) external returns (BalanceDelta) {
+        (bool success, bytes memory returnData) = address(implementation).delegatecall(
+            abi.encodeWithSelector(this.afterRemoveLiquidity.selector, sender, key, params, delta, hookData)
+        );
+        if (!success) {
+            revert FailedImplementationCall();
+        }
+        (bytes4 selector, BalanceDelta returnDelta) = abi.decode(returnData, (bytes4, BalanceDelta));
+        if (selector != BaseHook.afterRemoveLiquidity.selector) {
+            revert Hooks.InvalidHookResponse();
+        }
+        uint256 nonzeroDeltaCount = manager.getNonzeroDeltaCount();
+        if (nonzeroDeltaCount == 0 && returnDelta == BalanceDeltaLibrary.ZERO_DELTA) {
+            return returnDelta;
+        }
+        if (
+            returnDelta.amount0() > int256(uint256(int256(delta.amount0())) * maxFeeBips / MAX_BIPS)
+                || returnDelta.amount1() > int256(uint256(int256(delta.amount1())) * maxFeeBips / MAX_BIPS)
+        ) {
+            revert HookTookTooMuchFee();
+        }
+        // error on overflow
+        returnDelta - delta;
+        uint256 nonzeroHookDeltaCount;
+        int256 hookDelta = manager.currencyDelta(address(this), key.currency0);
+        if (hookDelta != 0) {
+            if (-hookDelta != returnDelta.amount0()) {
+                revert HookInvalidDeltasAfterRemove();
+            }
+            nonzeroHookDeltaCount++;
+            if (nonzeroHookDeltaCount == nonzeroDeltaCount) {
+                return returnDelta;
+            }
+        }
+        hookDelta = manager.currencyDelta(address(this), key.currency1);
+        if (hookDelta != 0) {
+            if (-hookDelta != returnDelta.amount1()) {
+                revert HookInvalidDeltasAfterRemove();
+            }
+            nonzeroHookDeltaCount++;
+            if (nonzeroHookDeltaCount == nonzeroDeltaCount) {
```

Comment:
is this not the same code twice? also Im confused what this is checking? put into helper to reduce code repetition